# Day 2: Dive!
from [AdventOfCode website]()

> Now, you need to figure out how to pilot this thing.
> 
> It seems like the submarine can take a series of commands like forward 1, down 2, or up 3:
> 
> - forward X increases the horizontal position by X units.
> - down X increases the depth by X units.
> - up X decreases the depth by X units.
> 
> Note that since you're on a submarine, down and up affect your depth, and so they have the opposite result of what you might expect.
> 
> The submarine seems to already have a planned course (your puzzle input). You should probably figure out where it's going. For example:
> ```
> forward 5
> down 5
> forward 8
> up 3
> down 8
> forward 2
> ```
> Your horizontal position and depth both start at 0. The steps above would then modify them as follows:
> 
> - forward 5 adds 5 to your horizontal position, a total of 5.
> - down 5 adds 5 to your depth, resulting in a value of 5.
> - forward 8 adds 8 to your horizontal position, a total of 13.
> - up 3 decreases your depth by 3, resulting in a value of 2.
> - down 8 adds 8 to your depth, resulting in a value of 10.
> - forward 2 adds 2 to your horizontal position, a total of 15.
>
> After following these instructions, you would have a horizontal position of 15 and a depth of 10. (Multiplying these together produces 150.)
> 
> Calculate the horizontal position and depth you would have after following the planned course.
>
> **What do you get if you multiply your final horizontal position by your final depth?**

In [1]:
# Preparing some helper stuff
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns 
import numpy as np

# Let's read that file
dataFrame = pd.read_csv("input_day2.txt")
dataFrame.head()

Values
0  forward 2
1     down 2
2  forward 6
3  forward 8
4     down 8

In [2]:
# Convert from frame, extract directions as strings and movements as numbers
texts = dataFrame.Values.tolist()
directions = []
movements = []

# Values are stored as "forward 2", split at space
for text in texts:
    split = text.split(" ")
    directions.append(split[0])
    movements.append(int(split[1]))

# Get of list from that stuff
tuples = zip(directions, movements)
tuplesList = list(tuples)
tuplesList[0:4] # Check some

[('forward', 2), ('down', 2), ('forward', 6), ('forward', 8)]

In [3]:
# Compute what's required
pos = 0
depth = 0
for tuple in tuplesList[0:]:
    dir = tuple[0]
    val = tuple[1]    

    if dir == 'forward':
        pos = pos + val
    elif dir == 'down':
        depth = depth + val
    elif dir == 'up':
        depth = depth - val
    else:
        "Error !"

print(depth)
print(pos)
print(f"Multiplied value is {pos * depth}")

702
2034
Multipled value is 1427868


## Solution of task 1
The value searched is then **1427868**

# Task 2 

> Based on your calculations, the planned course doesn't seem to make any sense. You find the submarine manual and discover that the process is actually slightly more complicated.
> 
> In addition to horizontal position and depth, you'll also need to track a third value, aim, which also starts at 0. The commands also mean something entirely different than you first thought:
> 
> - down X increases your aim by X units.
> - up X decreases your aim by X units.
> - forward X does two things:
>     - It increases your horizontal position by X units.
>     - It increases your depth by your aim multiplied by X.
>     - Again note that since you're on a submarine, down and up do the opposite of what you might expect: "down" means aiming in the positive direction.
> 
> Now, the above example does something different:
> 
> - forward 5 adds 5 to your horizontal position, a total of 5. Because your aim is 0, your depth does not change.
> - down 5 adds 5 to your aim, resulting in a value of 5.
> - forward 8 adds 8 to your horizontal position, a total of 13. Because your aim is 5, your depth increases by 8*5=40.
> - up 3 decreases your aim by 3, resulting in a value of 2.
> - down 8 adds 8 to your aim, resulting in a value of 10.
> - forward 2 adds 2 to your horizontal position, a total of 15. Because your aim is 10, your depth increases by 2*10=20 to a total of 60.
> 
> After following these new instructions, you would have a horizontal position of 15 and a depth of 60. (Multiplying these produces 900.)
> 
> Using this new interpretation of the commands, calculate the horizontal position and depth you would have after following the planned course. What do you get if you multiply your final horizontal position by 
> your final depth?

In [5]:
# Compute what's required
pos = 0
depth = 0
aim = 0

for tuple in tuplesList[0:]:
    dir = tuple[0]
    val = tuple[1]    

    if dir == 'forward':
        pos = pos + val
        depth = depth + val * aim
    elif dir == 'down':
        aim = aim + val
    elif dir == 'up':
        aim = aim - val
    else:
        "Error !"

print(depth)
print(pos)
print(f"Multiplied value is {pos * depth}")

770963
2034
Multipled value is 1568138742


## Second task answer
Yiaaaa, the correct answer is then **1568138742**